### PR TITLE
Fix #796 and #807 : Kick method might only kick player on same map instance

### LINKED ISSUE #796: Miniland privacy change bug
**Describe the bug**
Changing Miniland privacy mode to "Private" or "Locked" cause respawn for other players on whole channel.

**To Reproduce**
1. Open 2 or more game windows with different characters
2. Change Miniland privacy mode to "Private" or "Locked"
3. ???
4. Opps.

**Expected behavior**
Miniland privacy mode changed, no one respawned.

**Demo**
TgarDrv1Rk

### PR DESCRIPTION
**Pull request by Bari77**
Could be more efficient to add unit tests for MapInstance.

### CODE REVIEW

File: src/NosCore.GameObject/Providers/MinilandProvider/MinilandProvider.cs [hunk 1]
```diff
@@ -155,6 +155,8 @@ public void SetState(long characterId, MinilandState state)
                 var ml = _minilandIds[characterId];
                 var miniland = _mapInstanceProvider.GetMapInstance(ml.MapInstanceId);
                 ml.State = state;
+
+                // TODO : If MinilandState.Private => Don't kick friends
```
Comment (0Lucifer0) [inline src/NosCore.GameObject/Providers/MinilandProvider/MinilandProvider.cs]: Codacy Issue found: [Complete the task associated to this 'TODO' comment.](https://app.codacy.com/app/NosCoreIO/NosCore/pullRequest?prid=4513002)

**Comment by 0Lucifer0** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/app/NosCore/NosCore/pullRequest?prid=4513002&bid=15205996)

**Comment by 0Lucifer0** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- src/NosCore.Controllers/Friend/MJoinPacketHandler.cs  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/NosCore/NosCore/pullRequest?prid=4513002&bid=15205996)

File: src/NosCore.Controllers/Friend/MJoinPacketHandler.cs [hunk 2]
```diff
@@ -45,20 +45,31 @@ public MJoinPacketHandler(IFriendHttpClient friendHttpClient, IMinilandProvider
...
+                if (miniland.State == MinilandState.Open)
                 {
-                    session.ChangeMapInstance(info.MapInstanceId, 5, 8);
+                    session.ChangeMapInstance(miniland.MapInstanceId, 5, 8);
                 }
                 else
                 {
+                    if (miniland.State == MinilandState.Private)
+                    {
+                        if (friendList.Where(w => w.RelationType != CharacterRelationType.Blocked)
```
Comment (0Lucifer0) [inline src/NosCore.Controllers/Friend/MJoinPacketHandler.cs]: Codacy Issue found: [Merge this if statement with the enclosing one.](https://app.codacy.com/app/NosCoreIO/NosCore/pullRequest?prid=4513002)

File: src/NosCore.Controllers/Miniland/MlobjPacketHandler.cs [hunk 3]
```diff
@@ -32,6 +33,7 @@ namespace NosCore.PacketHandlers.Miniland
     public class MlEditPacketHandler : PacketHandler<MLEditPacket>, IWorldPacketHandler
     {
         private readonly IMinilandProvider _minilandProvider;
+        private readonly IFriendHttpClient _friendHttpClient;
```
Comment (0Lucifer0) [inline src/NosCore.Controllers/Miniland/MlobjPacketHandler.cs]: Codacy Issue found: [Remove the unused private field '_friendHttpClient'.](https://app.codacy.com/app/NosCoreIO/NosCore/pullRequest?prid=4513002)

**Comment by 0Lucifer0** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/NosCore.Controllers/Friend/MJoinPacketHandler.cs  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/NosCore/NosCore/pullRequest?prid=4513002&bid=15205996)

File: test/NosCore.Tests/PacketHandlerTests/MJoinPacketHandlerTests.cs [hunk 4]
```diff
@@ -169,5 +169,79 @@ public void Join()
...
+                    IsConnected = true,
+                    CharacterName = _targetSession.Character.Name,
+                    RelationType = CharacterRelationType.Friend
+                }
+            });
+            _minilandProvider.Setup(s => s.GetMiniland(It.IsAny<long>())).Returns(new Miniland
+            { MapInstanceId = TestHelpers.Instance.MinilandId, State = MinilandState.Private });
+            _mjoinPacketHandler.Execute(mjoinPacket, _session);
+
+            var lastpacket = (InfoPacket)_session.LastPackets.FirstOrDefault(s => s is InfoPacket);
```
Comment (0Lucifer0) [inline test/NosCore.Tests/PacketHandlerTests/MJoinPacketHandlerTests.cs]: Codacy Issue found: [Remove the unused local variable 'lastpacket'.](https://app.codacy.com/app/NosCoreIO/NosCore/pullRequest?prid=4513002)

**Comment by 0Lucifer0** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- src/NosCore.Controllers/Friend/MJoinPacketHandler.cs  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/NosCore/NosCore/pullRequest?prid=4513002&bid=15205996)

**Comment by 0Lucifer0** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/NosCore.Controllers/Friend/MJoinPacketHandler.cs  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/NosCore/NosCore/pullRequest?prid=4513002&bid=15205996)

File: src/NosCore.GameObject/Networking/ClientSession/ClientSession.cs [hunk 5]
```diff
@@ -497,5 +498,10 @@ public void HandlePackets(IEnumerable<IPacket> packetConcatenated, IChannelHandl
                 }
             }
         }
+
+        public List<CharacterRelationStatus> GetFriends()
+        {
```
Comment (0Lucifer0) [inline src/NosCore.GameObject/Networking/ClientSession/ClientSession.cs]: Let’s not add stuff into this class.

File: src/NosCore.GameObject/Providers/MinilandProvider/MinilandProvider.cs [hunk 6]
```diff
@@ -148,16 +150,28 @@ public Miniland Initialize(Character character)
             return minilandInfo;
         }
 
-        public void SetState(long characterId, MinilandState state)
+        public void SetState(long characterId, MinilandState state, ClientSession clientSession)
```
Comment (0Lucifer0) [inline src/NosCore.GameObject/Providers/MinilandProvider/MinilandProvider.cs]: This method definition should not be changed

File: src/NosCore.GameObject/Providers/MinilandProvider/MinilandProvider.cs [hunk 7]
```diff
@@ -148,16 +150,28 @@ public Miniland Initialize(Character character)
...
                 var ml = _minilandIds[characterId];
                 var miniland = _mapInstanceProvider.GetMapInstance(ml.MapInstanceId);
                 ml.State = state;
+
                 if (ml.State != MinilandState.Open)
                 {
-                    miniland.Kick(o => o.VisualId != characterId);
+                    if (ml.State == MinilandState.Private)
+                    {
+                        List<long> friends = clientSession.GetFriends()
```
Comment (0Lucifer0) [inline src/NosCore.GameObject/Providers/MinilandProvider/MinilandProvider.cs]: Use friend service

File: src/NosCore.GameObject/Providers/MinilandProvider/IMinilandProvider.cs [hunk 8]
```diff
@@ -33,6 +35,6 @@ public interface IMinilandProvider
         List<Portal> GetMinilandPortals(long characterId);
         Miniland GetMinilandFromMapInstanceId(Guid mapInstanceId);
         void AddMinilandObject(MapDesignObject mapObject, long characterId, InventoryItemInstance minilandobject);
-        void SetState(long characterId, MinilandState state);
+        void SetState(long characterId, MinilandState state, ClientSession clientSession);
```
Comment (0Lucifer0) [inline src/NosCore.GameObject/Providers/MinilandProvider/IMinilandProvider.cs]: No

**Comment by 0Lucifer0** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/NosCore.Controllers/Friend/MJoinPacketHandler.cs  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/NosCore/NosCore/pullRequest?prid=4513002&bid=15205996)

File: src/NosCore.GameObject/Providers/MinilandProvider/IMinilandProvider.cs [hunk 9]
```diff
@@ -20,6 +20,8 @@
 using System;
 using System.Collections.Generic;
 using ChickenAPI.Packets.Enumerations;
+using NosCore.GameObject.HttpClients.FriendHttpClient;
+using NosCore.GameObject.Networking.ClientSession;
```
Comment (erwan-joly) [inline src/NosCore.GameObject/Providers/MinilandProvider/IMinilandProvider.cs]: useless

File: src/NosCore.GameObject/Providers/MinilandProvider/MinilandProvider.cs [hunk 10]
```diff
@@ -1,4 +1,4 @@
-﻿//  __  _  __    __   ___ __  ___ ___
+//  __  _  __    __   ___ __  ___ ___
```
Comment (erwan-joly) [inline src/NosCore.GameObject/Providers/MinilandProvider/MinilandProvider.cs]: ?

File: src/NosCore.GameObject/Providers/MinilandProvider/MinilandProvider.cs [hunk 11]
```diff
@@ -155,9 +159,21 @@ public void SetState(long characterId, MinilandState state)
...
                 var miniland = _mapInstanceProvider.GetMapInstance(ml.MapInstanceId);
                 ml.State = state;
+
                 if (ml.State != MinilandState.Open)
                 {
-                    miniland.Kick(o => o.VisualId != characterId);
+                    if (ml.State == MinilandState.Private)
+                    {
+                        List<long> friends = _friendHttpClient.GetListFriends(characterId)
+                            .Where(w => w.RelationType != CharacterRelationType.Blocked)
```
Comment (erwan-joly) [inline src/NosCore.GameObject/Providers/MinilandProvider/MinilandProvider.cs]: should be by default as friend and not blacklist

File: src/NosCore.GameObject/Providers/MinilandProvider/MinilandProvider.cs (hunk 10 above)
Comment (Bari77) [reply src/NosCore.GameObject/Providers/MinilandProvider/MinilandProvider.cs]: I don't know why this get changed, It's the first line with ASCI NOSCORE.

File: src/NosCore.GameObject/Providers/MinilandProvider/IMinilandProvider.cs (hunk 9 above)
Comment (Bari77) [reply src/NosCore.GameObject/Providers/MinilandProvider/IMinilandProvider.cs]: True, i modify it.
Forgot to run Resharper.

File: src/NosCore.GameObject/Providers/MinilandProvider/MinilandProvider.cs (hunk 11 above)
Comment (Bari77) [reply src/NosCore.GameObject/Providers/MinilandProvider/MinilandProvider.cs]: Hum, on CharacterEntityEntension, method GenerateFinit() line 148, we need to load entire FriendList no ? For load blocked relation on client.
I could by default said non blocked, and do the Where clause on GetListFriends method and add a boolean on it called "includeBlocked" by default at false, and apply Where clause on it ? Like that ?
`List<CharacterRelationStatus> GetListFriends(long visualEntityVisualId, boolean includeBlocked = false) {
    includeBlocked ? return Get<List<CharacterRelationStatus>>(visualEntityVisualId) : return Get<List<CharacterRelationStatus>>(visualEntityVisualId).Where(w => w.RelationType != CharacterRelationType.Blocked);
}`

Comment (erwan-joly) [reply src/NosCore.GameObject/Providers/MinilandProvider/MinilandProvider.cs]: the service never return blacklist in friends

Comment (Bari77) [reply src/NosCore.GameObject/Providers/MinilandProvider/MinilandProvider.cs]: Okay, i change it so ;) thx for reviewed it.

**Comment by 0Lucifer0** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/NosCore.Controllers/Friend/MJoinPacketHandler.cs  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/NosCore/NosCore/pullRequest?prid=4513002&bid=15205996)

File: src/NosCore.GameObject/Providers/MinilandProvider/MinilandProvider.cs [hunk 12]
```diff
@@ -155,9 +157,20 @@ public void SetState(long characterId, MinilandState state)
...
+
                 if (ml.State != MinilandState.Open)
                 {
-                    miniland.Kick(o => o.VisualId != characterId);
+                    if (ml.State == MinilandState.Private)
+                    {
+                        List<long> friends = _friendHttpClient.GetListFriends(characterId)
+                            .Select(s => s.CharacterId)
+                            .ToList();
+                        // Kick all players in miniland except owner and his friends
```
Comment (erwan-joly) [inline src/NosCore.GameObject/Providers/MinilandProvider/MinilandProvider.cs]: this comment is not really useful. try having comments only when really nescessary

**Pull request merged**